### PR TITLE
mds: throttle scrub start for multiple active MDS

### PR DESCRIPTION
**Pull request by mchangir**
- add check to "scrub start" command handler to dishonor request if multiple MDS are "active" for a recursive scrub request
- add check to MDSRank::handle_mds_map() to see if scrubbing is in progress and max_mds is greater than 1, if so then scrubbing is aborted. This is supposed to take care of the condition when scrubbing has been started with max_mds == 1 and then bumped to a higher value

Fixes: https://tracker.ceph.com/issues/43483
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by joscollin** [comment]
I had a quick look. Put that check in `command_scrub_start` that already iterates `scrubop_vec`?

**Comment by batrick** [comment]
@mchangir please see the discussion on IRC, in #ceph-devel.

**Comment by mchangir** [comment]
> I had a quick look. Put that check in `command_scrub_start` that already iterates `scrubop_vec`?

I put the check there since some error message could be returned to the user.
Since the command handler `command_scrub_start()` probably gets executed asynchronously by the `finisher`, the command would return quietly and the only thing that could be done would be to log the error condition to the logs. The user would then have to check the logs.
Asking the user to check the logs would not provide a good User Experience.

File: src/mds/MDSRank.cc [hunk 1]
```diff
@@ -2384,6 +2384,18 @@ void MDSRankDispatcher::handle_mds_map(
   if (mdsmap->get_inline_data_enabled() && !oldmap.get_inline_data_enabled())
     dout(0) << "WARNING: inline_data support has been deprecated and will be removed in a future release" << dendl;
 
+  if (scrubstack->is_scrubbing() && mdsmap->get_max_mds() > 1) {
+    std::lock_guard l(mds_lock);
```
Comment (batrick) [inline src/mds/MDSRank.cc]: This method already executes with the `mds_lock` locked. See `MDSDaemon::ms_dispatch2`.

File: src/mds/MDSRank.cc [hunk 2]
```diff
@@ -2384,6 +2384,18 @@ void MDSRankDispatcher::handle_mds_map(
...
+    C_SaferCond cond;
+    scrubstack->scrub_abort(&cond);
```
Comment (batrick) [inline src/mds/MDSRank.cc]: Use `C_MDSInternalNoop` instead.

File: src/mds/MDSRank.cc (hunk 1 above)
Comment (mchangir) [reply src/mds/MDSRank.cc]: > This method already executes with the `mds_lock` locked. See `MDSDaemon::ms_dispatch2`.

removed the locking

File: src/mds/MDSRank.cc (hunk 2 above)
Comment (mchangir) [reply src/mds/MDSRank.cc]: > Use `C_MDSInternalNoop` instead.

now used C_MDSInternalNoop

**Comment by batrick** [comment]
@mchangir I've made some adjustments to your PR as another commit. Please squash the commits when you get back online.

The scrub failure you found was unrelated.

File: src/mds/MDSRank.cc [hunk 3]
```diff
@@ -2385,11 +2385,9 @@ void MDSRankDispatcher::handle_mds_map(
     dout(0) << "WARNING: inline_data support has been deprecated and will be removed in a future release" << dendl;
 
   if (scrubstack->is_scrubbing()) {
-    unsigned active = mdsmap->get_num_mds(MDSMap::DaemonState::STATE_ACTIVE);
-    unsigned stopping = mdsmap->get_num_mds(MDSMap::DaemonState::STATE_STOPPING);
-    if (active + stopping > 1) {
-      C_MDSInternalNoop cond;
-      scrubstack->scrub_abort(&cond);
+    if (mdsmap->get_max_mds() > 1) {
```
Comment (mchangir) [inline src/mds/MDSRank.cc]: @batrick I had indeed used `mdsmap->get_max_mds()` earlier ... but that led to a backtrace dump in the logs when testing, especially when scrub is started with max_mds == 1 and then the max_mds is bumped to 2.

File: src/mds/MDSRank.cc [hunk 4]
```diff
@@ -2526,15 +2524,12 @@ void MDSRankDispatcher::handle_asok_command(
...
     /* if there are more than one mds active and a recursive scrub is requested,
      * dishonor the request
      */
-    unsigned active = mdsmap->get_num_mds(MDSMap::DaemonState::STATE_ACTIVE);
-    unsigned stopping = mdsmap->get_num_mds(MDSMap::DaemonState::STATE_STOPPING);
-    bool is_recursive = std::find(scrubop_vec.begin(), scrubop_vec.end(), string("recursive")) != scrubop_vec.end();
-    // if (mdsmap->get_max_mds() > 1 &&
-    if ((active + stopping > 1) && is_recursive) {
-      ss << "There are more than one MDS active. Hence a recursive scrub "
+    if (mdsmap->get_max_mds() > 1) {
```
Comment (mchangir) [inline src/mds/MDSRank.cc]: @batrick Don't we need to test for the `recursive` flag here ?

File: src/mds/MDSRank.cc [hunk 5]
```diff
@@ -2385,11 +2385,9 @@ void MDSRankDispatcher::handle_mds_map(
     dout(0) << "WARNING: inline_data support has been deprecated and will be removed in a future release" << dendl;
 
   if (scrubstack->is_scrubbing()) {
-    unsigned active = mdsmap->get_num_mds(MDSMap::DaemonState::STATE_ACTIVE);
-    unsigned stopping = mdsmap->get_num_mds(MDSMap::DaemonState::STATE_STOPPING);
-    if (active + stopping > 1) {
-      C_MDSInternalNoop cond;
-      scrubstack->scrub_abort(&cond);
+    if (mdsmap->get_max_mds() > 1) {
+      auto c = new C_MDSInternalNoop;
```
Comment (mchangir) [inline src/mds/MDSRank.cc]: @batrick Won't this be a memory leak ?

File: src/mds/MDSRank.cc (hunk 3 above)
Comment (batrick) [reply src/mds/MDSRank.cc]: The condition variable on the stack would explain that.

File: src/mds/MDSRank.cc (hunk 4 above)
Comment (batrick) [reply src/mds/MDSRank.cc]: No. Consider the directory may be fragmented and other fragments are on a different MDS.

File: src/mds/MDSRank.cc (hunk 5 above)
Comment (batrick) [reply src/mds/MDSRank.cc]: No, when the context is finished it destroys itself. See `src/include/context.h`.

**Pull request merged**